lens-data/lens/private/syntax/stx.rkt

```diff
@@ -36,7 +36,19 @@
 ;;                       [Any Any -> Any])
 (define (restore-stx stx dat)
   (if (syntax? stx)
-      (datum->syntax stx dat stx stx)
+      ;; Preserve the distinction between #'(a . (b c)) and #'(a b c)
+      (let loop ([stx1 stx]
+                 [dat1 dat])
+        (cond
+          [(syntax? stx1)
+           (datum->syntax stx1 (loop (syntax-e stx1) dat1) stx1 stx1)]
+          [(and (pair? stx1) (pair? dat1))
```

Comment:
Just pushed all the changes. I tend to split `(cons a b)` on two lines to make the two parts stand out, but for something this short it's not really warranted, so I switched to one line for condition and one for result.